homeassistant/components/google_assistant_sdk/__init__.py

```diff
@@ -0,0 +1,100 @@
+"""Support for Google Assistant SDK."""
+from __future__ import annotations
+
+import aiohttp
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, ConfigEntryState
+from homeassistant.const import CONF_NAME, CONF_TOKEN, Platform
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers import discovery
+from homeassistant.helpers.config_entry_oauth2_flow import (
+    OAuth2Session,
+    async_get_config_entry_implementation,
+)
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DEFAULT_ACCESS, DOMAIN
+from .helpers import async_send_text_commands
+
+SERVICE_SEND_TEXT_COMMAND = "send_text_command"
+SERVICE_SEND_TEXT_COMMAND_FIELD_COMMAND = "command"
+SERVICE_SEND_TEXT_COMMAND_SCHEMA = vol.All(
+    {
+        vol.Required(SERVICE_SEND_TEXT_COMMAND_FIELD_COMMAND): vol.All(
+            str, vol.Length(min=1)
+        ),
+    },
+)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Google Assistant SDK component."""
+    hass.async_create_task(
+        discovery.async_load_platform(
+            hass, Platform.NOTIFY, DOMAIN, {CONF_NAME: DOMAIN}, config
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Google Assistant SDK from a config entry."""
+    implementation = await async_get_config_entry_implementation(hass, entry)
+    session = OAuth2Session(hass, entry, implementation)
+    try:
+        await session.async_ensure_token_valid()
+    except aiohttp.ClientResponseError as err:
+        if 400 <= err.status < 500:
+            raise ConfigEntryAuthFailed(
+                "OAuth session is not valid, reauth required"
+            ) from err
+        raise ConfigEntryNotReady from err
+    except aiohttp.ClientError as err:
+        raise ConfigEntryNotReady from err
+
+    if not async_entry_has_scopes(hass, entry):
```

Comment:
I would suggest that you can exclude this required scopes check. This was originally added for google calendar because it had both a "readwrite" and a "read" scope and it needed to support re-auth if you changed the options settings to know to ask you to re-auth. Google sheets copied it, but I don't think it really needs this easier given it had one scope last I looked. 

I think you should omit this, then add back later if you introduce new scopes that need permissions.

Reply:
I see. However I don't see harm in keeping it. It validates the token has the proper scope in case someone manually modified the config text files.

Reply:
In google calendar it has had bugs before, which is why I am weary of this. It's also a "virus" as people keep copying it around so I feel like it should just be removed given the unnecessary complex, extra test coverage needed etc and I don't think people should continue to copy this.

Reply:
Done.